astropy/nddata/nddata.py

```diff
@@ -186,30 +187,70 @@ def __init__(
             elif data.meta is not None:
                 meta = data.meta
 
+            # get the data attribute as it is, and continue to process it:
             data = data.data
 
-        else:
-            if hasattr(data, "mask") and hasattr(data, "data"):
-                # Separating data and mask
+        # if the data is wrapped by astropy.utils.masked.Masked:
+        if isinstance(data, Masked):
+            # first get the mask if one is available:
+            if hasattr(data, "mask"):
                 if mask is not None:
                     log.info(
-                        "overwriting Masked Objects's current mask with specified mask."
+                        "overwriting Masked Quantity's current mask with specified mask."
                     )
                 else:
                     mask = data.mask
 
-                # Just save the data for further processing, we could be given
-                # a masked Quantity or something else entirely. Better to check
-                # it first.
-                data = data.data
+            if isinstance(data, MaskedNDArray):
+                if unit is not None and hasattr(data, "unit") and data.unit != unit:
+                    log.info(
+                        "overwriting MaskedNDArray's current unit with specified unit."
+                    )
+                    data = data.to(unit).value
+                elif unit is None and hasattr(data, "unit"):
+                    unit = data.unit
+                    data = data.value
+
+                # now get the unmasked ndarray:
+                data = np.asarray(data)
 
             if isinstance(data, Quantity):
-                if unit is not None and unit != data.unit:
+                # this is a Quantity:
+                if unit is not None and data.unit != unit:
                     log.info("overwriting Quantity's current unit with specified unit.")
-                else:
+                    data = data.to(unit)
+                elif unit is None and data.unit is not None:
                     unit = data.unit
                 data = data.value
 
+        if isinstance(data, np.ma.masked_array):
+            if mask is not None:
+                log.info(
+                    "overwriting masked ndarray's current mask with specified mask."
+                )
+            else:
+                mask = data.mask
+            data = data.data
+
+        if isinstance(data, Quantity):
+            # this is a Quantity:
+            if unit is not None and data.unit != unit:
+                log.info("overwriting Quantity's current unit with specified unit.")
+                data = data.to(unit)
+            elif unit is None and data.unit is not None:
+                unit = data.unit
+            data = data.value
+
+        if isinstance(data, np.ndarray):
+            # check for mask from np.ma.masked_ndarray
+            if hasattr(data, "mask"):
```

Comment:
Is this redundant since you check for `np.ma.masked_array` above at https://github.com/astropy/astropy/pull/14175/files#diff-6ee41247b537e2a969874c2122e1931966c666a6b09e891fffc1fbeae49dfac7R226

Reply:
This is necessary for Masked `Quantity` objects because: 
```python
import numpy as np
import astropy.units as u
from astropy.utils.masked import Masked

mq = Masked(u.Quantity(1, u.km), mask=True)

isinstance(mq, np.ndarray), isinstance(mq, np.ma.masked_array)
# returns (True, False)
```